File: fib_ts/fib_solution.ts
class Fibonacci {
  iterativeFib(num: number): number {
    let a = 0;
    let b = 1;

    for (let _ = 0; _ < num; _++) {
      [a, b] = [b, a + b];
    }
    return a;
  }

  recursiveFib(num: number): number {
    if (num === 0 || num === 1) {
      return num;
    }
    return this.recursiveFib(num - 1) + this.recursiveFib(num - 2);
  }

  dynamicFib(num: number): number {
    const cache: { [key: number]: number } = {};
    return this._dynamicFib(cache, num);
  }

  _dynamicFib(cache: { [key: number]: number }, num: number) {
    if (num === 0 || num === 1) {
      return num;
    }
    if (num in cache) {
      return cache[num];
    }
    cache[num] =
      this._dynamicFib(cache, num - 1) + this._dynamicFib(cache, num - 2);

    return cache[num];
  }
}

const fib = new Fibonacci();
console.log(fib.iterativeFib(7));
console.log(fib.recursiveFib(7));
console.log(fib.dynamicFib(7));

File: fib_py/fib_solution.py
class Fibonacci:
    def iterative_fib(self, num):
        a, b = 0, 1
        for _ in range(num):
            a, b = b, a + b
        return a

    def recursive_fib(self, num):
        if num == 0 or num == 1:
            return num
        return self.recursive_fib(num - 1) + self.recursive_fib(num - 2)

    def dynamic_fib(self, num):
        cache = {}
        return self._dynamic_fib(cache, num)

    def _dynamic_fib(self, cache, num):
        if num == 0 or num == 1:
            return num
        if num in cache:
            return cache[num]
        cache[num] = self._dynamic_fib(cache, num - 1) + self._dynamic_fib(
            cache, num - 2
        )
        return cache[num]


fib = Fibonacci()
print(fib.iterative_fib(7))
print(fib.recursive_fib(7))
print(fib.dynamic_fib(7))
